partiql-eval/src/eval/mod.rs

```diff
@@ -140,33 +144,52 @@ pub struct Evaluated {
 }
 
 /// Represents an evaluation context that is used during evaluation of a plan.
-pub trait EvalContext {
-    fn bindings(&self) -> &dyn Bindings<Value>;
+pub trait EvalContext<'a>: SessionContext<'a> + Debug {
+    fn as_session(&'a self) -> &'a dyn SessionContext<'a>;
+
     fn add_error(&self, error: EvaluationError);
     fn has_errors(&self) -> bool;
     fn errors(&self) -> Vec<EvaluationError>;
 }
 
-#[derive(Default, Debug)]
-pub struct BasicContext {
-    bindings: MapBindings<Value>,
-    errors: RefCell<Vec<EvaluationError>>,
+#[derive(Debug)]
+pub struct BasicContext<'a> {
+    pub bindings: MapBindings<Value>,
+
+    pub sys: SystemContext,
+    pub user: HashMap<String, &'a (dyn Any)>, // TODO: Unicase the keys?
```

Comment:
We'll probably define our own context type, but for the most flexibility, this should be `&'a (dyn Any + 'a)` to support types that contain lifetimes themselves (e.g. `&'a Foo<'a>` - else you can only store something that is `&'a Foo<'static>`).

...This doesn't seem to just work however.  I think the `user_context` method on the SessionContext trait needs to be updated as well, there might be some other hoops that need to be jumped through as well.

Reply:
`Any` is defined as 
```rust
pub trait Any: 'static {...}
```
(See https://doc.rust-lang.org/std/any/trait.Any.html)

So you can't add a `+ 'a`

Reply:
Ok, this won't work for us then.  Do you have some time to talk about this in person tomorrow?